doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2310 @@
+```rust
+match expr {
+  <some_complex_match_arm> => true,
+  _ => false,
+}
+// ... can be replaced with ...
+matches!(expr, some_complex_match_arm)
+```
+`matches!` supports multi-match and match guards as well.
+
+Irrefutable patterns can be used with normal variable declaration.
+The syntax `let x = ...;` actually uses a pattern: `x` is a pattern.
+Other irrefutable patterns can be used there:
+```rust
+let (a, b) = ...;  // Destructure a tuple, rather than using clunky `.0` and `.1` field names.
+let Foo { foo, bar, baz } = ...;  // Destructure a struct, to access its fields directly.
+let 42 = ...;  // Not allowed: `42` is not an irrefutable pattern.
+```
+
+Special variants of `if` and `while` exist to take advantage of patterns, too:
+```rust
+if let Some(x) = my_option {
+  // If the pattern succeeds, the body will be executed, and `x` will be bound
+  // to the value inside the Option.
+  do_thing(x);
+} else {
+  // Else block is optional; `x` is undefined here.
+  // do_thing(x);  // Error.
+}
+
+while let Some(x) = some_func() {
+  // Loop terminates once the pattern match fails. Again, `x` is bound
+  // to the value inside the Option.
+}
+```
+Unlike normal `let` statements, `if let` and `while let` expressions are meant to be used with refutable patterns.
+
+In general, almost every place where a value is bound can be an irrefutable pattern, such as function parameters and `for` loop variables:
+```rust
+fn get_first((x, _): (u32, u32)) -> u32 { x }
+
+for (k, v) in my_key_values { .. }
+```
+
+### Traits {#traits}
+
+Traits are Rust's core code-reuse abstraction.
+Rust traits are like interfaces in other languages: a list of methods that a type must implement.
+Traits themselves, however, are _not_ types.
+
+A very simple trait from the standard library is `Clone`[^88]:
+```rust
+trait Clone {
+  fn clone(&self) -> Self;
+}
+```
+
+A type satisfying `Clone`'s interface (in Rust parlance, "implements `Clone`") has a `clone` method with the given signature, which returns a duplicate of `self`.
+To implement a trait, you use a slightly funny `impl` syntax[^89]:
+```rust
+impl Clone for MyType {
+  fn clone(&self) -> Self { /* implementation */ }
+}
+```
+
+This gives us a consistent way to spell "I want a duplicate of this value".
+The standard library provides traits for a number of similar operations, such as `Default`[^90], for providing a default value, `PartialEq`[^91] and `Eq`, for equality, `PartialOrd`[^92] and `Ord`, for ordering, and `Hash`[^93], for non-cryptographic hashing.
+
+The above traits are special in that they have trivial implementations for a struct or enum, assuming that all fields of that struct or enum implement it.
+The `#[derive()]` syntax described in the "Ownership" section can be used with any of these traits to automatically implement them for a type.
+It is not uncommon for POD types to look like this:
+```rust
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct MyPod { ... }
```

Comment:
```suggestion
struct MyPod {
    a: u8,
    b: u8,
    // x: OtherCustomStruct,    <-- The derive would fail as the compiler cannot automatically implement `Copy` and related functions for non-POD types 
}
```

Reply:
Done.